File: frontend/src/types/conversation.ts
export interface Conversation {
  id: string;
  connectionId: string;
  title: string;
  createdAt: string;
  updatedAt: string;
}

export interface Message {
  id: number;
  conversationId: string;
  role: 'user' | 'assistant';
  content: string;
  toolCalls: ToolCallData[] | null;
  createdAt: string;
}

export interface ToolCallData {
  id: string;
  tool: string;
  input: Record<string, unknown>;
  status: 'running' | 'completed' | 'error';
  output: string | null;
  durationMs: number | null;
  precedingText?: string;
}

export interface ConversationWithMessages extends Conversation {
  messages: Message[];
}

export interface ConversationListResponse {
  items: Conversation[];
  total: number;
}

export interface CreateConversationRequest {
  title?: string;
}

export interface UpdateConversationRequest {
  title: string;
}

export interface CreateMessageRequest {
  role: 'user' | 'assistant';
  content: string;
  toolCalls?: ToolCallData[] | null;
}

export interface GenerateTitleRequest {
  firstMessage: string;
}

export interface GenerateTitleResponse {
  title: string;
}

File: backend/app/models/conversation.py
"""Pydantic models for agent conversations and messages."""

from datetime import datetime
from typing import Any

from humps import camelize
from pydantic import BaseModel, Field


def to_camel(string: str) -> str:
    return camelize(string)


class ConversationBase(BaseModel):
    model_config = {"alias_generator": to_camel, "populate_by_name": True}


class ConversationCreate(ConversationBase):
    title: str | None = Field(default=None, max_length=200)


class ConversationUpdate(ConversationBase):
    title: str = Field(..., min_length=1, max_length=200)


class ConversationResponse(ConversationBase):
    id: str
    connection_id: str
    title: str
    created_at: datetime
    updated_at: datetime


class MessageCreate(ConversationBase):
    role: str = Field(..., pattern="^(user|assistant)$")
    content: str
    tool_calls: list[dict[str, Any]] | None = None


class MessageResponse(ConversationBase):
    id: int
    conversation_id: str
    role: str
    content: str
    tool_calls: list[dict[str, Any]] | None = None
    created_at: datetime


class ConversationWithMessages(ConversationResponse):
    messages: list[MessageResponse] = []


class ConversationListResponse(ConversationBase):
    items: list[ConversationResponse]
    total: int


class GenerateTitleRequest(ConversationBase):
    first_message: str = Field(..., max_length=1000)


class GenerateTitleResponse(ConversationBase):
    title: str
